lang/cs/GrpcCodeGeneration.targets

```diff
@@ -0,0 +1,138 @@
+          string grpcOutputDirectory = Path.GetFullPath(OutputDirectory);
+          string grpcPluginName = Path.Combine(GrpcBinaryDirectory, "grpc_csharp_plugin.exe");
+
+          // Get the list of names of .proto files
+          string[] sourceFiles = SourceFileList.Split(new char[]{';'});
+
+          string copyright = string.Join(Environment.NewLine,
+            "// Licensed to the Apache Software Foundation (ASF) under one",
+            "// or more contributor license agreements.  See the NOTICE file",
+            "// distributed with this work for additional information",
+            "// regarding copyright ownership.  The ASF licenses this file",
+            "// to you under the Apache License, Version 2.0 (the",
+            "// \"License\"); you may not use this file except in compliance",
+            "// with the License.  You may obtain a copy of the License at",
+            "//",
+            "//   http://www.apache.org/licenses/LICENSE-2.0",
+            "//",
+            "// Unless required by applicable law or agreed to in writing,",
+            "// software distributed under the License is distributed on an",
+            "// \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY",
+            "// KIND, either express or implied.  See the License for the",
+            "// specific language governing permissions and limitations",
+            "// under the License.");
+
+
+          // Setup the conversion process.
+          ProcessStartInfo rProcInfo = new ProcessStartInfo()
+          {
+              RedirectStandardOutput = true,
+              RedirectStandardError = true,
+              UseShellExecute = false,
+              CreateNoWindow = true,
+              FileName = Path.Combine(grpcToolsDirectory ,"protoc.exe")
+          };
+
+          bool conversionError = false;
+          using (StreamWriter logWriter = new StreamWriter(Path.Combine(tempDir, "GrpcCodeGeneration.log")))
+          {
+              foreach (string file in sourceFiles)
+              {
+                  logWriter.WriteLine("Processing file: {0}", file);
+
+                  string fullFileName = Path.Combine(grpcSrcFileDirectory, file);
+                  rProcInfo.Arguments = " -I" + grpcSrcFileDirectory + " --csharp_out " + tempDir
+                      + " " + fullFileName + " --grpc_out " + tempDir
+                      + " --plugin=protoc-gen-grpc=" + grpcPluginName;
+                  logWriter.WriteLine("Command: {0}", rProcInfo.Arguments);
+
+                  if (File.Exists(fullFileName))
+                  {
+                      StringBuilder stdOutBuilder = new StringBuilder();
+                      StringBuilder stdErrBuilder = new StringBuilder();
+
+                      // Start the conversion process
+                      using (Process rProc = Process.Start(rProcInfo))
+                      {
+                          stdOutBuilder.Append(rProc.StandardOutput.ReadToEnd());
```

Comment:
Shouldn't WaitForExit come first? Then capture the stdout and stderr? Otherwise I believe you are capturing the current state of the stream, then waiting for exit.

Reply:
Good catch. What I wrote works for reading either only one of the standard out or standard error and is the recommended way. However, reading both requires tasks to prevent deadlock when either the stdout or stderr buffers are full.

See documentation here.
https://msdn.microsoft.com/en-us/library/system.diagnostics.processstartinfo.redirectstandardoutput(v=vs.110).aspx

Reply:
Done.